Replace old datasync walkthrough with the latest one (#33)
integr8ly/solution-patterns: walkthroughs/5-exploring-datasync-codeready/walkthrough.adoc

Currently Postgres and MongoDB are supported.
. Review your `./client/src/graphql/` folder containing client side queries for your data. These queries are automatically generated based on the server schema, and are used within the client application. Generating the client side queries helps developers get their client apps up and running quicker and helps them stay up to date as the server schema evolves.

=== Running DataSync client and server applications

. Open a new terminal window
. Execute the `prepare client` command in the new terminal. Client side application will be build and started. This can take a couple of minutes.
. Execute the `start server` command. This command starts the GraphQL server which also serves the client application for simplicity.
. The application should be opened in a preview window after build is finished.

[type=verification]
****
. Check if the website was loaded properly
. Select the + icon to create a new item
. On the new screen enter a `name` and `description` and create the task.
. New task should appear in the task list.

> ----

Reviewer: Is there supposed to be any code snippet? It creates this empty code snipped.
![image](https://user-images.githubusercontent.com/20477661/84382564-1ec34780-abeb-11ea-98ab-ded9100f4c50.png)
```suggestion
```

Author: I have noticed it. It was working in old version of the solution explorer so it might that format changed

// update the component versions for each release
:rhmi-version: 1

// URLs
:openshift-console-url: {openshift-host}/console
:sso-realm-url: {user-sso-url}/auth/admin/solution-patterns/console/index.html
:data-sync-documentation-url: https://access.redhat.com/documentation/en-us/red_hat_managed_integration/{rhmi-version}/html-single/developing_a_data_sync_app/index

//attributes
:integreatly-name: Managed Integration
:data-sync-name: Data Sync
:data-sync-starter: Data Sync Starter

> :customer-sso-name: SSO

Reviewer 2: Red Hat SSO?

Author: I think I will remove this as we not using that anyway

This solution pattern will show you how to:

* Build a DataSync server based on your business model
* Protect the application's frontend and backend using {customer-sso-name}.
* Explore all the capabilities provided by {data-sync-name}.

The following diagram shows the architecture of the {data-sync-starter}:

image::images/arch.png[architecture, role="integr8ly-img-responsive"]

[type=walkthroughResource, serviceName=openshift]
.Red Hat OpenShift
****
* link:{openshift-console-url}[Console, window="_blank"]
* link:https://docs.openshift.com/dedicated/4/welcome/index.html[OpenShift Documentation, window="_blank"]
* link:https://blog.openshift.com/[OpenShift Blog, window="_blank"]
****

[type=walkthroughResource]
.Data Sync
****
* link:{data-sync-documentation-url}[Getting Started with {data-sync-name}, window="_blank"]
****

:sectnums:

[time=15]
== Creating your DataSync project using DataSync Starter template

{data-sync-name} allows you to focus on your business model by giving you the ability
to generate a fully functional GraphQL based API and Node.js Server and Client Side components.

Developers can use GraphQL types to define the models their application deals with

> and generate the underlying backend that works out of the box with the RHMI services like SSO or AMQ Online.

Reviewer 2: ```suggestion
and generate the underlying backend that works out of the box with the RHMI services like Red Hat-SSO or AMQ Online.
```

Using GraphQL subscriptions, your application can receive live updates thanks to GraphQL and the DataSync client allows your application to operate independently of the network connection. 

> DataSync-Starter is an application template that gives developers 

Reviewer 2: ```suggestion
DataSync-Starter is an application template that gives developers an
```

opiniated implementation for React Ionic PWA and Mobile application backend by Node.js Server.

> DataSync as framework consist of the two major components:

Reviewer 2: Should this be `DataSync Framework`, or do you mean `DataSync as a framework`?
```suggestion
DataSync Framework consist of the two major components:
```

* DataSync client (upstream https://offix.dev) 
+
Can be added to any web or mobile application

> to provide DataSynchronization capabilities

Reviewer 2: ```suggestion
to provide Data synchronization capabilities
```

> * Node.js Server with code generation capabilties (upstream https://graphback.dev) 

Reviewer 2: No longer code generation capabilities (unless you are talking about client code)

```suggestion
* Node.js Server with out of the box GraphQL API that is backed by MongoDB (upstream https://graphback.dev) 
```

+

> Server is providing ability to dynamically generate Node.js GraphQL server with DataSynchronization and Conflict resolution capabilties 

Reviewer 2: Not one word:

```suggestion
Server is providing ability to dynamically generate Node.js GraphQL server with Data synchronization and Conflict resolution capabilities 
```

. Navigate to the solution explorer
. Go to the "All Services" tab 
. Go to "Online IDE" and select "Open Console"
. Login to the CodeReady Workspaces dashboard
. Choose `NodeJS MongoDB Web Application` from list of the available stacks
. Enter `https://github.com/aerogear/datasync-starter#walkthrough` for the repository URL
. In the right corner press down arrow and select `Create and Start Editing`
. In the new window select the `Dev File` tab and the copy dev file from 
https://raw.githubusercontent.com/aerogear/datasync-starter/master/devfile.yaml
. Press `Create`
. Wait for the workspace to start
. Press `Open` to open CodeReady editor
. In the active workspace, run the `Install dependencies` command.
+
The commands are listed in the `MyWorkspace` view available by selecting the box icon in the right top corner of the screen.

=== Creating your DataSync model

Based on a user defined **Model** DataSync generates the underlying GraphQL API and database resources.
The starter template uses MongoDB as the default datasource for storing application data.

. In your project explorer view on the left,  go to the `./model/task.graphql` file. 

> This file contains an link:https://graphql.org/learn/schema/#object-types-and-fields[GraphQL Schema, window="_blank"]. Here you can define types of the data your application will work with. Graphback will use your model and generate underlying data access methods in the form of GraphQL Queries and Mutations

Reviewer 2: ```suggestion
This file contains a link:https://graphql.org/learn/schema/#object-types-and-fields[GraphQL Schema, window="_blank"]. Here you can define types of the data your application will work with. Graphback will use your model and generate underlying data access methods in the form of GraphQL Queries and Mutations
```

> . Please add a `@datasync` annotation under the `@model` annotation on the `Task` type.

Reviewer 2: I believe this is different now, `@delta`?

Author: Yes. This is basing on the master branch. We have renamed that to delta to avoid confusion, but I will be able to make this change only after we release 0.14.0 Graphback

Those annotations control various behaviour of DataSync.
`@model` will create standard data access methods, while `@datasync` will provide data synchronization capabilities.
. Models can be changed by adding new types or fields. Add a new field to the `Task` type by adding `address: String`
. Your task model should be as follows
----
""" 
@model
@datasync
"""
type Task {
  id: ID!
  title: String!
  description: String!
  status: TaskStatus
  address: String
}
----

> === Generating your DataSync Node.JS server and React App

Reviewer 2: ```suggestion
=== Generating your DataSync Node.js server and React App
```

* `./client` - contains a React application written in TypeScript. `client/src/index.tsx` is the entrypoint to the client application.
* `.graphqlrc.yml` - 

. Review the `.graphqlrc.yml` file. This is a config file that
** determines what types of CRUD operations will be generated into GraphQL Queries and Mutations in your application.
** enables and configures various plugins that are used during the code generation process.
. Make sure that all fields in the `crud` section are enabled 
. Execute graphback cli command to generate source code:
`yarn graphback generate`. You can also execute it as predefined `generate source code` command in CodeReady
. Review `./server/src/schema/schema.qraphql`. 
This file has the original model and it also contains generated Queries and Mutations. The types of queries and mutations included are based on the `crud` fields in `.graphqlrc.yml`.
. Review the generated resolver files in `./server/src/resolvers/resolvers.ts`
This file contains methods used to fetch and modify data. Each individual method uses a 

> preconfigured `MongoDataProvider` which is an abstraction over a MongoDB client. Developers can point resolvers to any datasource.

Reviewer 2: MongoDBDataProvider
```suggestion
preconfigured `MongoDBDataProvider` which is an abstraction over a MongoDB client. Developers can point resolvers to any datasource.
```

****

[type=verificationFail]
****
Check the logs of the console
Verify that you followed each step in the procedure above.  
If you are still having issues, contact your administrator.
****

=== Interacting with the GraphQL Playground

The GraphQL Playground is an in browser GraphQL IDE that lets you directly perform queries and mutations against your GraphQL API.
It's a convenient way to interact with your GraphQL API without using a client application.
It is served directly by your server application as a developer tool and can be disabled in production.
In this section we will focus on using the playground.

. Open a new terminal window
. Execute `yarn start:server`
. Open the GraphQL Playground URL printed in console.
You can use the GraphQL playground to interact with the server API as described in the next step.
. Go to the Playground interface and replace the text in the left pane of the screen with the following query and mutation:

----
query listTasks {
  findAllTasks {
    title,
    description,
    address,
    id
  }
}

mutation createTask {
  createTask(input: {title: "complete the walkthrough", description: "complete the GraphQL walkthrough", address: "NA"}) {

>     title,

Reviewer 2: I don't think it will work with commas?
